google/cloud/bigtable/internal/async_retry_unary_rpc_test.cc

```diff
@@ -0,0 +1,250 @@
+// Copyright 2018 Google LLC.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/bigtable/internal/async_retry_unary_rpc.h"
+#include "google/cloud/bigtable/testing/mock_completion_queue.h"
+#include "google/cloud/bigtable/testing/mock_mutate_rows_reader.h"
+#include "google/cloud/testing_util/assert_ok.h"
+#include "google/cloud/testing_util/chrono_literals.h"
+#include <google/bigtable/admin/v2/bigtable_table_admin.grpc.pb.h>
+#include <gmock/gmock.h>
+#include <thread>
+
+namespace google {
+namespace cloud {
+namespace bigtable {
+inline namespace BIGTABLE_CLIENT_NS {
+namespace internal {
+namespace {
+
+namespace btadmin = ::google::bigtable::admin::v2;
+using ::testing::_;
+using ::testing::Invoke;
+
+class MockClient {
+ public:
+  MOCK_METHOD3(
+      AsyncGetTable,
+      std::unique_ptr<grpc::ClientAsyncResponseReaderInterface<btadmin::Table>>(
+          grpc::ClientContext*, btadmin::GetTableRequest const&,
+          grpc::CompletionQueue* cq));
+
+  MOCK_METHOD3(AsyncDeleteTable,
+               std::unique_ptr<grpc::ClientAsyncResponseReaderInterface<
+                   google::protobuf::Empty>>(grpc::ClientContext*,
+                                             btadmin::DeleteTableRequest const&,
+                                             grpc::CompletionQueue* cq));
+};
+
+TEST(AsyncRetryUnaryRpcTest, ImmediatelySucceeds) {
+  using namespace google::cloud::testing_util::chrono_literals;
+
+  MockClient client;
+
+  using ReaderType =
+      ::google::cloud::bigtable::testing::MockAsyncResponseReader<
+          btadmin::Table>;
+  auto reader = google::cloud::internal::make_unique<ReaderType>();
+  EXPECT_CALL(*reader, Finish(_, _, _))
+      .WillOnce(Invoke([](btadmin::Table* table, grpc::Status* status, void*) {
+        // Initialize a value to make sure it is carried all the way back to
+        // the caller.
+        table->set_name("fake/table/name/response");
+        *status = grpc::Status(grpc::StatusCode::OK, "mocked-status");
```

Comment:
What role does `"mocked-status"` play?

Reply:
None, removed.

Reply:
There are a bunch of these in other tests.  I'm happy to use my fledgling skills to clean those up if you like.

Reply:
Thanks!  You may want to wait until the `noex::` classes get removed, because many tests will also get removed at the same time.  But feel free if you do not mind that your cleanup gets removed in a couple of weeks.

Reply:
I'll take a closer look to see if I can just avoid the `noex::` cases.  Thanks.